dvipdfm/dvipdfm.1
.TH dvipdfm 1 5/28/2001
.SH NAME
dvipdfm \- Produce
.SM PDF files directly from
.SM DVI
files

.SH SYNOPSIS
.B dvipdfm
.RI [ options ]
.IR file [.dvi]

.SH DESCRIPTION
The program
.B dvipdfm
generates a
.SM PDF
file from a
.SM DVI
file.
.SM DVI
files are the output format
produced by TeX.
Output from 
.B groff
can be converted to
.SM DVI
format
using
.BR grodvi .

The program
.B dvipdfm
recognizes several commonly used
.BR \especial
commands, which are extensions to the
.SM DVI
format.
Specifically, it understands
color specials, papersize specials,
tpic specials (which allow it to be used with
.BR gpic ),
hypertex specials, and some PostScript
specials.  These extensions allow documents
to contain color, figures, and hyperlinks.
The program tries to mimic
the behavior of
.B dvips
so that many
macro packages produced for use with 
.B dvips
will work with 
.B dvipdfm.
Dvipdfm understands its own specific
.B \especial
commands to allow access to PDF features
such as annotations and bookmarks.
For more details, consult the 
.I "Dvipdfm User's Manual,"
which is distributed with
.B dvipdfm.

.SH OPTIONS

.TP 5
.B \-\^c
Ignore color
.BR \especials .
The
.B \-\^c
toggles a flag that disables color
.B \especial
processing.
Unless changed in the configuration file,
that flag is off, and color
.BR \especial "s"
are interpreted normally.
The
.B \-\^c
option may be used to produce a black and white document
from a document containing color TeX
.B \especial
commands.

.TP 5
.B \-\^d
Remove thumbnail images after including them.
(See the
.B \-\^t
option for more information.)

.TP 5
.B \-\^e
Toggle partial font embedding flag.  If partial
font embedding is off, embedded fonts are
.I fully
embedded.
The default, if not changed in the configuration file,
is to embed only those glyphs actually used in the document.

.TP 5
.B \-\^f map_file
Set the name of the font map file to 
.IR map_file .
The format of the font map file
is documented in the 
.IR "Dvipdfm User's Manual" .

.TP 5
.B \-\^l
Select landscape mode.  In other words, exchange the 
.I x
and
.I y
dimensions of the paper.

.TP 5
.B \-\^m " mag"
Magnify the input document by
.IR mag .

.TP 5
.B \-\^o " name"
Generate PDF output file having the name
.IR name .
By default, the name of the output
file is
.IR file .pdf.

.TP 5
.B \-\^p " paper"
Select the papersize by name (e.g.,
.BR letter ", " legal ", " ledger ", " tabloid ", " a3 ", " a4 ", or " a5
)

.TP 5
.B \-\^r " size"
Set resolution of bitmapped fonts to
.B size
dots per inch.  Bitmapped fonts are generated
by the Kpathsea library, which uses MetaFont.  Bitmapped
fonts are included as type 3 fonts in the PDF output file.

.TP 5
.B \-\^s " page_specifications"
Select the pages of the
.SM DVI
file to be converted.
The
.I page_specifications
consists of a comma separated list of
.I page_ranges:

.IR "page_specifications := page_specification" "[," page_specifications "]"

where

.IR "page_specification := single_page" | page_range

.IR "page_range := " [ first_page ] \^\-\^[ last_page ]

An empty
.I first_page
is implied to be the first page of the
.SM DVI
file.
An empty
.I last_page
is treated as the last page of the
.SM DVI
file
(e.g.,
.BR "" `` "\-\^s 1,3,5" '' includes
pages 1, 3, and 5; 
.BR "" `` "\-\^s \-" ''
includes all pages;
.BR "" `` "\-\^s \-,\-" ''
includes two copies of all pages in the
.SM DVI
file;
and
.BR `` "\-\^s 1\^\-\^10" ''
includes the first ten pages of the
.SM DVI
file.)

.TP 5
.B \-\^t
Search for thumbnail images of each page
in the directory named by the
.B TMP
environment variable.
The thumbnail images must be named in a specific format.
They must have the same base name
as the
.SM DVI
file and they must have the page number as the extension
to the file name.  Dvipdfm does not
generate the thumbnails itself, but it is distributed with a wrapper
program named 
.B dvipdft 
that does so.

.TP 5
.B \-\^v 
Increase verbosity.
Results of the 
.B \-\^v
option are cumulative (e.g., 
.B \-\^vv
) increases the verbosity by two increments.

.TP 5
.B \-\^x x_offset
Set the left margin to 
.I x_offset
.  The default left margin is
.BR 1.0in .
The dimension may be specified in any units understood by TeX (e.g.,
.BR bpt ", " pt ", " in ", " cm
)

.TP 5
.B \-\^y y_offset
Set the top margin to 
.I y_offset
.  The default top margin is
.BR 1.0in .
The dimension may be specified in any units understood by TeX (e.g.,
.BR bpt ", " pt ", " in ", " cm
)

.TP 5
.B \-\^z compression_level
Set the compression level to 
.IR compression_level .
Compressions levels range from 0 (no compression) to
9 (maximum compression) and correspond to the values understood by zlib.

.SH ENVIRONMENT

.B Dvipdfm 
uses the
.B kpathsea
library for locating the files that it opens.  Hence,
the environment variables documented in the
.I Kpathsea library
documentation influence
.B dvipdfm.

Dvipdfm also reads the following environment variables:

.TP 5
TMP
The directory to search for thumbnail images of each page.

.SH FILES
The precise location of the following files is determined by the
.I Kpathsea library
configuration.  The location may be determined by using kpsewhich, e.g.,
.nf
.B kpsewhich \-progname=dvipdfm \-format='other text files' config
.fi

.TP 5
.I config
Startup options

.TP 5
.I t1fonts.map
The default font map file (this name may be changed in the config file).

.TP 5
.I *.tfm
TeX font metrics

.TP 5
.I *.vf
TeX virtual font files

.TP 5
.I *.pfb
PostScript Type 1 font files

.TP 5
.I 
texmf.cnf
The Kpathsea library configuration file.
The location of this file may be found by typing
.nf
.B kpsewhich texmf.cnf
.fi


.SH "SEE ALSO"
.BR tex "(1), " dvips "(1),"
and the
.I "Kpathsea library"
info documentation.

.SH AUTHOR
Mark A. Wicks



dvipdfm/NEWS

   WARNING: THE LATEX SUPPORT FILE NAMED DVIPDFM.DEF HAS CHANGED IN VERSION
	    0.12 TO BE MORE COMPATIBLE WITH DVIPS AND OTHER DVI DRIVERS.  IT
	    CANNOT BE USED WITH EARLIER VERSIONS OF DVIDFPM. HOWEVER,
	    EARLIER VERSIONS OF DVIPDFM.DEF CAN BE USED WITH VERSION 0.12.

September 6, 1999 - Version 0.12

   Added native support for including MetaPost output.  Along with this
   comes the ability to interpret simple PostScript operators natively.
   This allows dvipdfm to process DVI files intended for dvips as long as
   they don't do anything too fancy.  The internal mini-PostScript
   interpreter is good enough to process the "rotate" commands produced by
   the standard dvips graphics driver.

   Added support for PostScript files using a helper application such as
   GhostScript.  PostScript files may be included with either the "PSfile"
   special or the "plotfile" special.

   Added support for the "standard" color specials.  Together with the above
   list, dvipdfm can now process many DVI files targets for dvips (as long
   as they use very simple PostScript commands).  PostScript code such as
   that produced by PSTricks is not supported.

   Added a configuration file for specifying default command line options.

August 21, 1999.

   WARNING:  VERSION 0.11 CHANGES THE FORMAT OF THE t1fonts.map FILE.

   The font code has been reworked in several ways.  Version 0.11 does not
require AFM files.  This is a significant improvement since AFM files are
disappearing from major distributions and repositories.  Unfortunately, this
required a change in the format of the t1fonts.map file.

   A bit of the font code has been rewritten to provide work-alikes for the
"slantfont" and "extendfont" provided by dvips.  There are several
workarounds for various bugs in Acrobat Reader involving searching for and
selecting ligatures and there's also an attempt to address the disappearing
Gamma problem.

   Version 0.11 also adds PNG and thumbnail support.
   
   Version 0.11 add support for the HyperTeX (HTML) specials for improved
DVI file portability.
   
April 8, 1999.

   Version 0.10.5 changes the rule code to use a PDF stroke instead
of filling a rectangular box.  It also omits some unecessary
font information from the PDF file for the base 14 PS fonts.
This is to produce for slightly smaller PDF files when using
these fonts.

March 30, 1999.

   NOTE: VERSION 0.10.4 CHANGES THE NAME/LOCATION OF ITS CONFIGURATION FILES!

   It also fixes a bug that caused nested virtual fonts to fail and fixes a couple of other
   fairly insignificant bugs.

   Version 0.10.0 introduced PK fonts support and support
   for TPIC special commands to allow some native graphics
   support (don't you hate distilling those .eps files).

December 30, 1998.
   Version 0.9.0 supports partial font embedding and Flate compression. The
   page tree and destination tree are shipped out as balanced trees.

December 14, 1998.
   Version 0.8.3 fixes a memory leak in the vf code found by Akiro Kakuto.
   Another change hopefully makes running ./configure a bit easier.

   Version 0.8.1 and 0.8.2 fix an overflow in the kerning code
   which occurs for large horizontal motions.

   Version 0.8 supports virtual fonts.  Much of the code has been
   reorganized and rewritten.  It is nearly an order of magnitude faster than
   earlier versions.  Version 0.8 no longer requires the kpath source
   tree.  Instead it uses an installed version of libkpathsea and its
   headers.

Nov. 28, 1998
   Specifying font encoding now supported.  There is a font map file for
   the more common PostScript fonts.

Nov. 26, 1998

   LaTeX Graphics and hyperref support is now available. See the directory
   latex-support for information and drivers.


dvipdfm/dvipdfm.c
/*  $Header: /home/mwicks/Projects/Gaspra-projects/cvs2darcs/Repository-for-sourceforge/dvipdfm/dvipdfm.c,v 1.3 1998/11/29 21:37:16 mwicks Exp $

    This is dvipdf, a DVI to PDF translator.
    Copyright (C) 1998  by Mark A. Wicks

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
    
    The author may be contacted via the e-mail address

	mwicks@kettering.edu
*/

	
#include <stdio.h>
#include "c-auto.h"
#include "system.h"
#include "dvi.h"
#ifdef KPATHSEA
#include <kpathsea/progname.h>
#endif
#include "mem.h"
#include "string.h"
#include "pdfdoc.h"
#include "pdfdev.h"

#ifndef HAVE_BASENAME
const char *basename (const char *s);
#endif /* HAVE_BASENAME */

struct rect 
{
  double width;
  double height;
};
typedef struct rect rect;

struct 
{
  char *s;
  struct rect data;
} paper_sizes[] = {
  {"letter" , { 612.0, 792.0}},
  {"legal" , { 612.0, 1008.0}},
  {"ledger" , { 1224.0, 792.0}},
  {"tabloid" , { 792.0, 1224.0}},
  {"a4" , { 595.27, 841.82}},
  {"a3" , { 841.82, 1190.16}}};

static rect get_paper_size (char *string)
{
  int i;
  for (i=0; i<sizeof(paper_sizes)/sizeof(paper_sizes[0]); i++) {
    if (!strcmp (string, paper_sizes[i].s))
      break;
  }
  if (i == sizeof(paper_sizes)/sizeof(paper_sizes[0]))
    ERROR ("Paper size is invalid");
  return paper_sizes[i].data;
}


char *dvi_filename = NULL, *pdf_filename = NULL;

static void set_default_pdf_filename(void)
{
  const char *dvi_base;
  dvi_base = basename (dvi_filename);
  if (strlen (dvi_base) < 5 || strncmp (".dvi", dvi_base+strlen(dvi_base)-4, 4)) 
  {
    pdf_filename = NEW (strlen(dvi_base)+5, char);
    strcpy (pdf_filename, dvi_base);
    strcat (pdf_filename, ".pdf");
  } else
  {
    pdf_filename = NEW (strlen(dvi_base)+1, char);
    strncpy (pdf_filename, dvi_base, strlen(dvi_base)-4);
    strcpy (pdf_filename+strlen(dvi_base)-4, ".pdf");
  }
}

static void usage (void)
{
   fprintf (stderr, "%s, version %s, Copyright (C) 1998 by Mark A. Wicks\n", PACKAGE, VERSION);
   fprintf (stderr, "dvipdfm comes with ABSOLUTELY NO WARRANTY.\n");
   fprintf (stderr, "This is free software, and you are welcome to redistribute it\n");
   fprintf (stderr, "under certain conditions.  Details are distributed with the software.\n");
   fprintf (stderr, "\nUsage: dvipdf [options] dvifile\n");
   fprintf (stderr, "where [options] is one or more of\n\n");
   fprintf (stderr, "\t-o filename\tOutput file name\n");
   fprintf (stderr, "\t-p papersize\tpapersize is letter, legal, ledger, tabloid, a4, or a3\n");
   fprintf (stderr, "\t-l \t\tlandscape mode\n");
   fprintf (stderr, "\t-m filename\tFont map file name [pdffonts.map]\n");
   exit(1);
}

static double paper_width = 612.0, paper_height = 792.0;
static landscape_mode = 0;

static void do_args (int argc, char *argv[])
{
  while (argc > 0) {
    if (*argv[0] == '-') {
      switch (*(argv[0]+1)) {
      case 'o':  
	if (argc < 2)
	  ERROR ("Missing output file name");
	pdf_filename = NEW (strlen(argv[1])+1,char);
	strcpy (pdf_filename, argv[1]);
	argv += 2;
	argc -= 2;
	break;
      case 'p':
	{
	  rect paper_size = get_paper_size (argv[1]);
	  if (argc < 2)
	    ERROR ("Missing paper size");
	  paper_width = paper_size.width;
	  paper_height = paper_size.height;
	  argv += 2;
	  argc -= 2;
	}
	break;
      case 'l':
	landscape_mode = 1;
	argv += 1;
	argc -= 1;
	break;
      case 'm':
	type1_set_mapfile (argv[1]);
	argv += 2;
	argc -= 2;
	break;
      default:
	usage();
      }
    } else {
      if (dvi_filename != NULL) {
	fprintf (stderr, "Multiple filenames specified\n");
	exit(1);
      }
      if (strncmp (".dvi", argv[0]+strlen(argv[0])-4, 4)) {
	dvi_filename = NEW (strlen (argv[0])+1+4, char);
	strcpy (dvi_filename, argv[0]);
	strcat (dvi_filename, ".dvi");
      }
      else {
	dvi_filename = NEW (strlen (argv[0])+1, char);
	strcpy (dvi_filename, argv[0]);
      }
      argv += 1;
      argc -= 1;
    }
  }
  if (dvi_filename == NULL) {
    fprintf (stderr, "No dvi filename specified\n");
    exit(1);
  }
}


int CDECL main (int argc, char *argv[]) 
{
  int i;
  static int really_quiet = 0;
  if (argc < 2) {
    usage();
    return 1;
  }
#ifdef KPATHSEA
  kpse_set_progname (argv[0]);
#endif
  argv+=1;
  argc-=1;
  do_args (argc, argv);

  /* Check for ".dvi" at end of argument name */
  if (pdf_filename == NULL)
    set_default_pdf_filename();
  
  if (!really_quiet)
    fprintf (stdout, "%s -> %s\n", dvi_filename, pdf_filename);
  dvi_open (dvi_filename);
  dvi_init (pdf_filename);
  if (landscape_mode)
    dev_set_page_size (paper_height, paper_width);
  else
    dev_set_page_size (paper_width, paper_height);

  for (i=0; i<dvi_npages(); i++) {
    fprintf (stderr, "[%d", i+1);
    dvi_do_page (i);
    fprintf (stderr, "]");
  }
  dvi_complete ();
  dvi_close();
  fprintf (stderr, "\n");
  return 0;

}

